jax/_src/numpy/ufuncs.py

```diff
@@ -152,15 +152,25 @@ def _arccosh(x: ArrayLike, /) -> Array:
     out = _where(real(out) < 0, lax.neg(out), out)
   return out
 
+def _preserving_dtype_floor(x: ArrayLike, /) -> Array:
+    if dtypes.isdtype(dtypes.dtype(x), "integral"):
+      return lax.asarray(x)
+    return lax.floor(x)
+
+def _preserving_dtype_ceil(x: ArrayLike, /) -> Array:
+    if dtypes.isdtype(dtypes.dtype(x), "integral"):
+      return lax.asarray(x)
+    return lax.ceil(x)
+
 fabs = _one_to_one_unop(np.fabs, lax.abs, True)
 bitwise_invert = _one_to_one_unop(getattr(np, 'bitwise_invert', np.invert), lax.bitwise_not)
 bitwise_invert = _one_to_one_unop(getattr(np, 'bitwise_invert', np.invert), lax.bitwise_not)
 bitwise_not = _one_to_one_unop(np.bitwise_not, lax.bitwise_not)
 invert = _one_to_one_unop(np.invert, lax.bitwise_not)
 negative = _one_to_one_unop(np.negative, lax.neg)
 positive = _one_to_one_unop(np.positive, lambda x: lax.asarray(x))
-floor = _one_to_one_unop(np.floor, lax.floor, True)
-ceil = _one_to_one_unop(np.ceil, lax.ceil, True)
+floor = _one_to_one_unop(np.floor, _preserving_dtype_floor)
+ceil = _one_to_one_unop(np.ceil, _preserving_dtype_ceil)
```

Comment:
If this is no longer a one-to-one unary operation, we should not use this factory function: just define the function directly.

Reply:
@jakevdp sorry, I was confused about the implementation, how about this one https://github.com/google/jax/pull/21441/commits/ce2c9bb0bd6e6a9d49cb3de781bc5470084d7f96 ?

Reply:
Looks reasonable, except we should call `check_arraylike` on the inputs.